### PR TITLE
shutdown: remount /run without no-exec, before pivot root.

### PR DESCRIPTION
**Pull request by xnox**

### CODE REVIEW

File: src/core/shutdown.c [hunk 1]
```diff
@@ -146,6 +148,23 @@ static int parse_argv(int argc, char *argv[]) {
 }
 
 static int switch_root_initramfs(void) {
+        struct statfs st;
+        int r;
+
+        r = statfs("/run", &st);
+        if (r < 0)
+                log_error_errno(errno, "Failed to determine mount flags for /run: %m");
```
Comment (keszybz) [inline src/core/shutdown.c:156]: `log_warning` and "..., ignoring".

File: src/core/shutdown.c [hunk 2]
```diff
@@ -146,6 +148,23 @@ static int parse_argv(int argc, char *argv[]) {
...
+
+        /* If /run is mounted with NOEXEC, remount without it, as otherwise exec /shutdown post switch_root will fail */
+        if (r == 0 && (st.f_flags & ST_NOEXEC)) {
+                if (mac_smack_use())
+                        r = mount(NULL, "/run", NULL, MS_REMOUNT|MS_NOSUID|MS_NODEV|MS_STRICTATIME, "mode=755,smackfsroot=*");
+                else
+                        r = mount(NULL, "/run", NULL, MS_REMOUNT|MS_NOSUID|MS_NODEV|MS_STRICTATIME, "mode=755");
```
Comment (keszybz) [inline src/core/shutdown.c:163]: Shouldn't this be `st.f_flags & ~ST_NOEXEC | MS_REMOUNT`? Also, I'd prefer to use tertiary operator just for the third argument: `mac_smack_use() ? "mode=755,smackfsroot=*" : "mode=755".

File: src/core/shutdown.c [hunk 3]
```diff
@@ -146,6 +148,23 @@ static int parse_argv(int argc, char *argv[]) {
...
+                if (r < 0)
+                        return log_error_errno(errno, "Failed to remount /run without MS_NOEXEC: %m");
```
Comment (keszybz) [inline src/core/shutdown.c:165]: I don't think this should error out. Just print a warning and continue.

**Comment by poettering** [comment]
What's the rationale here? I mean /run is not noexec anyway, so why bother?

**Comment by poettering** [comment]
Also, when remounting the superblock we really need to use the original mount options I figure, we do that in umount.c and it's not pretty, but it's kinda needed, kernel file systems require that. See the commenting in mount_points_list_get() for more info

**Comment by mbiebl** [comment]
> What's the rationale here? I mean /run is not noexec anyway, so why bother?

I don't think this is true for all initramfs generators.

**Comment by poettering** [comment]
> > What's the rationale here? I mean /run is not noexec anyway, so why bother?
> 
> I don't think this is true for all initramfs generators.

 But that sounds like something to fix then, no?

File: src/core/shutdown.c (hunk 2 above)
Comment (xnox) [reply src/core/shutdown.c:163]: Ooooh, I like both of your suggestions!

File: src/core/shutdown.c (hunk 3 above)
Comment (xnox) [reply src/core/shutdown.c:165]: If we failed to remount, execv of /shutdown will fail as the almost next action. I guess we don't care, right..... ?

Comment (keszybz) [reply src/core/shutdown.c:165]: Yeah. The remounting is just an additional step, essentially a work-around for it being mounted wrong.

**Comment by keszybz** [comment]
I think we should be OK with mounting /run noexec. Even if we don't default to it upstream, we might want to some day. And there might be specialized distributions which have verified that they don't have anything which would need to exec from /run during normal runtime, and only during shutdown.

**Comment by poettering** [comment]
> I think we should be OK with mounting /run noexec. Even if we don't default to it upstream, we might want to some day. And there might be specialized distributions which have verified that they don't have anything which would need to exec from /run during normal runtime, and only during shutdown.

well, uh. I mean the XDG_RUNTIME_DIR specs for example (which are not precisely about /run, but about something below /run) explicitly declare that runtime dirs should be fully featured, and that means executables should be OK too. When I wrote that spec back then the idea was really that apps can drop in /run whatever they like, and should not expect a borked, reduced feature file system there... 

Now, /run is for system services what XDG_RUNTIME_DIR is for unpriv user stuff. It's kinda weird if the version for system services is less featured than the version for user code. In particular as there is system code that actually wants to place executable stuff there, which we know for a fact: initrds that want to allow transitioning back from the host OS... I mean, doing noexec outside of the initrd just appears like an excercise in complicating things needlessly, since you can#t start out with it, and you can shutdown with it, but only an inner window coudl support this...

Hence, I think we should probably go the other direction, and push downstreams to not place noexec needlessly on /run, in the interest of keeping interfaces similar between distros. We probably even should update file-hierarchy.7 to say that /run is expected to be "fully-featured", too, just like the XDG_RUNTIME_DIR spec says it to clarify this.

That all said, I am not going to block this PR. If people want this, it's OK by me... However it needs to be done properly, i.e. the mount option string should not be eaten up by the remount, it needs to be retained. Yeah, Linux API makes this hard when remounting the superblock, but it's how it is...

(That said, event though I don#t believe in the whole excercise: why does the initrd even insist on mounting the superblock MS_NOEXEC? I mean, if it really wants MS_NOEXEC to be set, it could just use it with MS_BIND, so that it only applies to the binding, not to the superblock. That would make it a ton easier to switch it on or off at arbitrary times since the mount option string doesn't matter anymore)

**Comment by xnox** [comment]
Hmmm... Currently initramfs-tools does
```
mount -t tmpfs -o "noexec,nosuid,size=${RUNSIZE:-10%},mode=0755" tmpfs /run
```
But using MS_NOEXEC flag would be nicer, however I'm not sure how that could be done from shell.

**Pull request closed without merging**